### ISSUE TITLE
Harden ssh_filter script

### ISSUE THREAD

**Issue by Faexa** [digint/btrbk#45]
Hi!
First of all, thank you for this wonderful program ! Very useful and handy !

I open this issue because I've some concerns about security when btrbk operates over ssh with the ssh_filter_btrbk script.

Let me explain you this:
- I have multiple source computers to be backuped to a target server
- If one computer gets compromised, the attacker cannot do whatever he wants thanks to the ssh_filter_btrbk script. But he can access dangerous btrfs command system-wide (root permissions).

For example:
1. Attacker can destroy other computers backups or system partitions (`btrfs subvolume delete` accepted for all subvolumes)
2. He can extract data from other backups or system (`btrfs sub snap` and `btrfs send`)

Maybe I'm paranoid and this kind of attack is highly unlikely but, I think It would be good to restrict the path in which the btrfs command can proceed and if out of this path: reject_and_die !

If you think this is relevant and needs more attention, I've made a modification of the ssh_filter_script to take into consideration the working path.
(as an example, my authorized_keys looks like this :
`from="192.168.0.1",command="/usr/local/bin/ssh_filter_btrbk.sh --log --sudo --restrict-path /btrfs/pool_1/backups/pc1"`)

I'm no programmer, and even less a bash expert... so my modifications are probably bad but
tell me what you think of all this !

- PS1: it uses `/bin/bash`, I don't know of to do this regular expression (with =~ operator) valid for all shells
- PS2: the regex is awful, but seems to be working (no regex expert neither ...): 
	- complicated because we can put multiple subvolume behind `btrfs sub delete`
	- So different subvolumes are separated with a space (since it's not authorized in the subvolumes name in btrbk, it souldn't be a problem) and the regex verifies that each space separated subvolume begins with the restricted path. And with the reject of `../` in the path, we should be good ?
- PS3: implemented `use_sudo` because I'm not connecting to remote root user and btrfs command needs these permissions

```bash
#!/bin/bash

set -e
set -u

export PATH=/sbin:/bin:/usr/sbin:/usr/bin

enable_log=
restrict_path=
use_sudo=


while [ "$#" -ge 1 ]
do
key="$1"

case $key in
    -l|--log)
    enable_log=1
    ;;

    -p|--restrict-path)
    restrict_path="$2"
    shift # past argument
    ;;

    -s|--sudo)
    use_sudo="sudo"
    ;;

    *)
    ;;
esac
shift

done



log_cmd()
{
    if [ -n "$enable_log" ]; then
        logger -p $1 -t ssh_filter_btrbk.sh "$2 (Name: ${LOGNAME:-<unknown>}; Remote: ${SSH_CLIENT:-<unknown>}; Restrict-path: $restrict_path; $use_sudo): $SSH_ORIGINAL_COMMAND"
    fi
}

reject_and_die()
{
    log_cmd "auth.err" "btrbk REJECT"
    /bin/echo "ERROR: ssh command rejected" 1>&2
    exit 1
}

run_cmd()
{
    log_cmd "auth.info" "btrbk ACCEPT"
    $use_sudo $SSH_ORIGINAL_COMMAND
}


check_path()
{
    if [ -n "$restrict_path" ]; then
	regex="^btrfs[-a-zA-Z0-9 ]* ([ ]*$restrict_path[^ ]*)*[ ]*$"

		if [[ $SSH_ORIGINAL_COMMAND =~ $regex ]]; then
			run_cmd
		else
			reject_and_die
		fi
    else
	run_cmd
    fi
}


case "$SSH_ORIGINAL_COMMAND" in
    *\$*) reject_and_die ;;
    *\&*) reject_and_die ;;
    *\(*) reject_and_die ;;
    *\{*) reject_and_die ;;
    *\;*) reject_and_die ;;
    *\<*) reject_and_die ;;
    *\>*) reject_and_die ;;
    *\`*) reject_and_die ;;
    *\|*) reject_and_die ;;
    *"../"*) reject_and_die;;		#prevent directory traversal
    btrfs\ subvolume\ show\ *)      check_path ;; #run_cmd ;;   # mandatory
    btrfs\ subvolume\ list\ *)      check_path ;; #run_cmd ;;   # mandatory
    btrfs\ subvolume\ snapshot\ *)  check_path ;; #run_cmd ;;   # mandatory if this host is backup source
    btrfs\ send\ *)                 check_path ;; #run_cmd ;;   # mandatory if this host is backup source
    btrfs\ receive\ *)              check_path ;; #run_cmd ;;   # mandatory if this host is backup target
    btrfs\ subvolume\ delete\ *)    check_path ;; #run_cmd ;;   # mandatory if scheduling is active
    btrfs\ subvolume\ find-new\ *)  check_path ;; #run_cmd ;;   # needed for "btrbk diff"
    btrfs\ filesystem\ usage\ *)    check_path ;; #run_cmd ;;   # needed for "btrbk info"
    *) reject_and_die ;;
esac
```

**Comment by digint** [comment]
Hi!
Thanks for your contribution, you opened an issue that I wanted to address since the beginning of btrbk development. The reasons why I never implemented this are:
- I wanted to keep the script as simple as possible
- Posix shell is (as far as I know) incapable of doing regular expressions reasonably
- There are people that don't have bash (for good reasons)
- Implementing proper security in bash is still a pain
- Writing bash is always a pain...

Well nevertheless, I took your proposed script, changed the regular expressions to be as readable as possible, and added some more options (you can now specify which type of access is allowed: source/target/delete/info, as well as specify multiple path restrictions using command-line options):

https://github.com/digint/btrbk/blob/harden_ssh_filter_btrbk/ssh_filter_btrbk.sh

example: `ssh_filter_btrbk.sh --log --target --delete --info -p /mnt/btr_pool/alpha -p /mnt/btr_pool/beta`
(Maybe the command-line option naming should be changed to something more intuitive.)

Let me know what you think, I guess I'll merge this into the next release as I think it's a big improvement, even if it relies on bash now.

**Comment by Faexa** [comment]
Nice improvement, the possibility of choosing which btrfs command can be launch and multiple path are good addition.

But I've encountered two bugs in the script :

1. The log_cmd raise an error : 
[not exact error : translated from french] 
`Bad substitution: no closing « } » in  ; Restrict-path: '/${restrict_path//\|/\', \'/ `
(what is this part supposed to do ? [not good in bash :) ]

  
2. The regular expression forces a trailing slash at the end with a filepath:

I've this btrbk configuration : 
```
volume /btrfs/pool_hdd
     snapshot_create  always
     subvolume home/alex
          snapshot_dir               home_snapshots/
          snapshot_name          alex.home
          target send-receive     ssh://nas/btrfs/pool_1/backups/alex-pc
```

So, `btrbk tree` and `btrbk info` send respectively these commands:

`btrfs subvolume show /btrfs/pool_1/backups/alex-pc`
`btrfs filesystem usage /btrfs/pool_1/backups/alex-pc`

Which are rejected because not matching the regex.
- I think the regular expression should be something like: 

```^(${btrfs_allow})( ${option_match})*( /(${restrict_path})(/|/${file_match})*)+$```
- and deleting the trailing slash added at line 31 : `restrict_path="${restrict_path}|${rpath}"`

In this case:
- the path can be the restricted path only without subvolume
  - without trailing slash at the end (like btrbk the commands show and list)
  - or with one optionally (like the receiving command: `btrfs receive /btrfs/pool_1/backups/alex-pc/`)

- the path can contain optionally a subvolume for deletion command for example

Is it good for you?
Thanks!

**Comment by digint** [comment]
> The log_cmd raise an error : [...]
oh, well I hacked this without really testing... guess I'll have to adapt the logging line anyways as it gets really long

Regarding the trailing slash: I basically added this in order to make sure that the restricted path is really a directory and not just a prefix, which I think is a good idea. The regex should read "use exact restrict-path or restrict-path-slash-somefile". Your regex fix above kind of fixes that too, but gets more unreadable again... I'll look into it again, this time with more "real-life" testing.

**Comment by digint** [comment]
Another try with no leading-slash-magic, seems pretty readable and correct to me. Also performed some successful tests:

https://github.com/digint/btrbk/blob/harden_ssh_filter_btrbk/ssh_filter_btrbk.sh

**Comment by Faexa** [comment]
Seems good to me !
I can't try it these next two days. I'll give you feedback on monday after testing.
Thank you for working this fast on the problem !

**Comment by Faexa** [comment]
The script seems to be working perfectly.
Command line options tested: log / sudo / restrict path (one path and two paths) / source / delete / info
All commands works correctly with and without trailing slash / with and without subvolume after the path / rejected when out of path, no directory traversal, rejected when path is used as a prefix / logging works
What was not tested: option source (and  snapshot / send) but I see no reason why this won't work !

Thank you for working on it !

**Comment by digint** [comment]
Thanks for the testing!

Merged into master: f01304df35fcfb4501d32caf47ed44ec09af7274
